Source/WebKit/UIProcess/API/Cocoa/WKDownloadDelegatePrivate.h

```diff
@@ -50,9 +50,25 @@ WK_SWIFT_UI_ACTOR
  @param download The download for which we need a placeholder policy
  @param completionHandler The completion handler that should be invoked with the chosen policy
  @discussion The placeholder policy specifies whether a placeholder file should be created in
- the Downloads directory when the download is in progress.
+ the Downloads directory when the download is in progress. If the client opts out of the
+ placeholder feature, it can choose to provide a custom URL to publish progress against.
+ This is useful if the client maintains it's own placeholder file.
  */
-- (void)_download:(WKDownload *)download decidePlaceholderPolicy:(void (^)(_WKPlaceholderPolicy))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
+- (void)_download:(WKDownload *)download decidePlaceholderPolicy:(void (^)(_WKPlaceholderPolicy, NSURL *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
+
+/* @abstract Called when the download receives a placeholder URL
+ @param download The download for which we received a placeholder URL
+ @param completionHandler The completion handler that should be called by the client in response to this call. 
+ @discussion The placeholder URL will normally refer to a file in the Downloads directory
+ */
+- (void)_download:(WKDownload *)download didReceivePlaceholderURL:(NSURL *)url completionHandler:(void (^)(void))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```

Comment:
Are these callbacks just informative, or do we actually need to wait for something to happen before continuing?

Reply:
Good point! The callback in didReceivePlaceholderURL  is used to send a message back to the Network process to inform that the client has received the placeholder, but the callback for didReceiveFinalURL is just informative, and can be removed.